scipy/linalg/_basic.py

```diff
@@ -141,11 +145,37 @@ def solve(a, b, lower=False, overwrite_a=False,
     array([ True,  True,  True], dtype=bool)
 
     """
+    xp = array_namespace(a, b)
+    if check_finite:
+        a = _asarray(a, check_finite=True, xp=xp)
+        b = _asarray(b, check_finite=True, xp=xp)
+    if is_numpy(xp):
+        return _solve(a, b, lower=lower, overwrite_a=overwrite_a,
+                      overwrite_b=overwrite_b, assume_a=assume_a, transposed=transposed)
+    unsupported_args = {
+        'lower': lower,
+        'assume_a': assume_a != 'gen',
+        'transposed': transposed
+    }
+    if any(unsupported_args.values()):
+        xp_unsupported_args(unsupported_args)
+    if hasattr(xp, 'linalg'):
+        dtype = xp.result_type(a, b, xp.float32)
+        a = xp.astype(a, dtype)
+        b = xp.astype(b, dtype)
```

Comment:
You might want to do this dtype work even if you have to fall back to NumPy. The reason is that if a package didn't have `linalg.solve` and then adds it, the behavior could (conceivably) change if it starts doing dtype conversions before the arrays passed into the function.

Reply:
that seems reasonable, I can't remember whether there was any reason for including this only behind this guard.